### PR TITLE
Revert "[JENKINS-70922] Remove Prototype `Ajax.Request` usage from `select.js`"

### PR DESCRIPTION
**Pull request by basil**
Reverting #7949 until an ATH regression is tracked down and fixed.

### Testing done

Before: https://ci.jenkins.io/job/Core/job/acceptance-test-harness/job/PR-1163/1/
After: https://ci.jenkins.io/job/Core/job/acceptance-test-harness/job/PR-1163/2/

### Proposed changelog entries

N/A (unreleased regression)

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7965"></a>

### CODE REVIEW

**Review by NotMyFault — APPROVED** [review]

**Review by timja — APPROVED** [review]
something is resetting credentials to their default value. Easy to reproduce in pipeline by trying to set a git credential to fetch the pipeline script.

I've stepped through it in the debugger and haven't managed to figure it out yet (even having it not reproduce when debugging -.-).

**Review by basil — COMMENTED** [review]
Expediting for tomorrow's weekly release.

**Comment by timja** [comment]
I've spent another hour or so on this and haven't figured it out yet.

The select gets called at least twice, first time has the value selected set, the second doesn't.

Any help appreciated

**Comment by basil** [comment]
This seems like a race condition around event firing. That is, the code was fragile to begin with but the race was not exposed; now, the move away from Prototype.js has perturbed the timing slightly and the race is exposed. You can see someone else has been here before in [JENKINS-33418](https://issues.jenkins.io/browse/JENKINS-33418). When we hit this race `selectEl.value` is empty by the time we get to `hasChanged`, which is not the case in the working version (where it is the new credential ID). But this is just evidence that something has gone wrong far earlier (unclear what).

You will likely need to spend more than an hour to solve this problem. In my experience, race conditions can rarely be debugged in a short amount of time. Interactive debugging will also be of little help. You might need to sprinkle the code with `console.log` statements to compare the execution before and after, not just in `select.js` but also in other code that creates and fires event handlers elsewhere in the codebase.

**Comment by basil** [comment]
@timja This seems to chase away the problem

```javascript
diff --git a/core/src/main/resources/lib/form/select/select.js b/core/src/main/resources/lib/form/select/select.js
index 315d0c4ace..082d376760 100644
--- a/core/src/main/resources/lib/form/select/select.js
+++ b/core/src/main/resources/lib/form/select/select.js
@@ -4,6 +4,7 @@ function updateListBox(listBox, url, config) {
   config = config || {};
   config = object(config);
   var originalOnSuccess = config.onSuccess;
+  var originalOnFailure = config.onFailure;
   var l = listBox;
 
   // Hacky function to retrofit compatibility with tables-to-divs
@@ -92,6 +93,9 @@ function updateListBox(listBox, url, config) {
           l.options[0] = null;
         }
       }
+      if (originalOnFailure !== undefined) {
+        originalOnFailure(rsp);
+      }
     });
   };
 
@@ -129,7 +133,15 @@ Behaviour.specify("SELECT.select", "select", 1000, function (e) {
   }
 
   // controls that this SELECT box depends on
-  refillOnChange(e, function (params) {
+  var onChange = function (params) {
+    if (e.hasAttribute("isRefilling")) {
+      // Still refilling; wait until the first refill operation is complete before beginning the second one.
+      setTimeout(function () {
+        onChange(params);
+      }, 100);
+      return;
+    }
+    e.setAttribute("isRefilling", true);
     var value = e.value;
     updateListBox(e, e.getAttribute("fillUrl"), {
       parameters: params,
@@ -153,7 +165,12 @@ Behaviour.specify("SELECT.select", "select", 1000, function (e) {
         if (hasChanged(e, value)) {
           fireEvent(e, "change");
         }
+        e.removeAttribute("isRefilling");
+      },
+      onFailure: function () {
+        e.removeAttribute("isRefilling");
       },
     });
-  });
+  };
+  refillOnChange(e, onChange);
 });
```

by serializing the filling activity, but it is neither a satisfying solution (because it does not get to the bottom of why the second change event was fired to begin with) nor a satisfying implementation (because of the 100-millisecond retrying).

**Comment by timja** [comment]
Thanks, tested locally and it looks good, raised: https://github.com/jenkinsci/jenkins/pull/7982
